Aggregations - support for partitioning set of terms

markharwood opened the pull request:
Filters terms into arbitrary sized partitions so that multiple requests can be done without trying to compute everything in one request.

First draft for review of approach - needs tests/docs etc

Closes #21487

jpountz commented on core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/support/IncludeExclude.java:
```diff
@@ -138,6 +169,22 @@ public boolean accept(BytesRef value) {
+                if (i % incNumPartitions == zeroBasedPartition){                
+                    acceptedGlobalOrdinals.set(i);
+                }
+            }
```
This is true. The alternative way I was thinking about would be to pull a terms enum from the sorted set doc values, and set bits in the bit set based on the hash of the values rather than their ordinal. This way, the partitioning would be based on the value but the terms aggregation would still be able to leverage ordinals to do the bucketing. The drawback is that it requires to compute a hash on every term of the field.

jimczi replied:
+1 for the terms enum pulled from the sorted set. It should be fast to enumerate the terms since it is a sequential reads in the term dict of each segment. The hashing will be slow though, but again it's great if we can have a way to exhaust a term aggregations consistently even if it's not as fast as we would like ;).

markharwood replied:
On the 2.7m userID terms test I see ords taking 2.4s vs 2.8s using hashes of terms.

jpountz commented on core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/support/IncludeExclude.java:
```diff
@@ -138,6 +169,29 @@ public boolean accept(BytesRef value) {
+            final TermsEnum termEnum = globalOrdinals.termsEnum();
+            
+            BytesRef term = termEnum.next();
+            int ord = 0;
```
you need a long, or this could overflow. Or even not store it at all, and use `termsEnum.ord()` to know the ord of the current term

jpountz commented on core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/support/IncludeExclude.java:
```diff
@@ -138,6 +169,29 @@ public boolean accept(BytesRef value) {
+            int ord = 0;
+            while (term != null) {
+                int hash = Math.abs(term.hashCode());
+                if (hash % incNumPartitions == incZeroBasedPartition) {
```
I would use `Math.floorMod(term.hashCode(), incNumPartitions)` to avoid issues in the case that the hashcode is `Integer.MIN_VALUE`

jpountz commented on core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/support/IncludeExclude.java:
```diff
@@ -61,15 +65,35 @@
+        public boolean accept(long value) {
+            // hash the value to keep even distributions
+            buffer.putLong(0, value);
+            final int hashCode = StringHelper.murmurhash3_x86_32(buffer.array(), 0, Long.BYTES, StringHelper.GOOD_FAST_HASH_SEED);            
```
you could call `BitMixer.mix64(value)`, which should have the same quality but might be a bit faster

jpountz commented on core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/support/IncludeExclude.java:
```diff
@@ -283,6 +354,13 @@ public IncludeExclude(StreamInput in) throws IOException {
         } else {
             excludeValues = null;
         }
+        if (in.getVersion().after(Version.V_5_0_0)) {
```
I think the version should be 5_1_0

jpountz replied:
you can have a look at SimpleQueryStringBuilder.VERSION_5_1_0_UNRELEASED to see how to do it

jpountz commented on core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/support/IncludeExclude.java:
```diff
@@ -436,12 +518,22 @@ public boolean token(String currentFieldName, XContentParser.Token token, XConte
+                            if (parseFieldMatcher.match(currentFieldName, PARTITION_FIELD)) {
+                                otherOptions.put(INCLUDE_FIELD, parser.intValue());
+                            }      
                         }
```
can you add an else block that throws an exception?

jpountz commented on core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/support/IncludeExclude.java:
```diff
@@ -480,15 +572,39 @@ public boolean token(String currentFieldName, XContentParser.Token token, XConte
+                            throw new IllegalArgumentException(PARTITION_FIELD.getPreferredName()+" must be between >1 and <="+pOf);                                                    
+                        }
+                    } else {
+                        throw new IllegalArgumentException(OF_FIELD.getPreferredName()+" value must be an integer");                        
```
it might make sense to throw a different exception message if pOf is null?

jpountz commented on core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/support/IncludeExclude.java:
```diff
@@ -480,15 +572,39 @@ public boolean token(String currentFieldName, XContentParser.Token token, XConte
+                    if(ofObject instanceof Integer){
+                        pOf = (Integer) ofObject;
+                        if (pOf < 2){
+                            throw new IllegalArgumentException(PARTITION_FIELD.getPreferredName()+" must be >1");                                                    
```
Should we make the partitions 0-based like the `from` parameter of the search API?

markharwood replied:
If we do we should change the "of" parameter name. "partition 19 of 20" doesn't sound like the last partition any more.

jimczi replied:
+1 for the 0-based partitions, this would be consistent with the `from` and the `slice`.

markharwood replied:
OK - I'll change`of` to be `num_partitions` just because "19 _of_ 20" seems misleading

jpountz commented on core/src/test/java/org/elasticsearch/search/aggregations/bucket/DoubleTermsIT.java:
```diff
@@ -359,6 +362,32 @@ private void testIncludeExcludeResults(double[] includes, double[] excludes, dou
+                assertFalse(foundTerms.contains(bucket.getKeyAsNumber()));
+                foundTerms.add(bucket.getKeyAsNumber());
+            }
+        }
```
maybe we should also assert on the size of `foundTerms` to ensure we visited all of them?

jimczi replied:
+1

jimczi commented on core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/support/IncludeExclude.java:
```diff
@@ -480,15 +572,39 @@ public boolean token(String currentFieldName, XContentParser.Token token, XConte
+                            throw new IllegalArgumentException(PARTITION_FIELD.getPreferredName()+" must be >1");                                                    
+                        }
+                        if (partition <1 || partition > pOf){
+                            throw new IllegalArgumentException(PARTITION_FIELD.getPreferredName()+" must be between >1 and <="+pOf);                                                    
```
must be  >=1 ?

jimczi commented on core/src/test/java/org/elasticsearch/search/aggregations/bucket/DoubleTermsIT.java:
```diff
@@ -359,6 +362,32 @@ private void testIncludeExcludeResults(double[] includes, double[] excludes, dou
+    }
+    
+    private void runTestFieldWithPartitionedFiltering(String field) throws Exception {
+        final int numPartitions = randomIntBetween(1, 4);
```
maybe start at 2, numPartitions=1 should fail, no ?

jimczi commented on core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/support/IncludeExclude.java:
```diff
@@ -250,13 +307,27 @@ public IncludeExclude(long[] includeValues, long[] excludeValues) {
     }
 
+    public IncludeExclude(int partition, int pOf) {
+        this.incZeroBasedPartition = partition - 1;
```
Check that partition >= 1 && pof > partition ? It's only checked when we (de)serialize from json.

jimczi commented on core/src/test/java/org/elasticsearch/search/aggregations/bucket/LongTermsIT.java:
```diff
@@ -326,6 +329,37 @@ private void testIncludeExcludeResults(long[] includes, long[] excludes, long[]
+    }
+    
+    private void runTestFieldWithPartitionedFiltering(String field) throws Exception {
+        final int numPartitions = randomIntBetween(1, 4);
```
same here, randomIntBetween(2, 4) ?

markharwood commented:
Thanks for the review comments. 
One thing left which is the 5.0.1 version check which I'll address when I rebase to latest master before the final push so I can see that constant.

jpountz commented:
> One thing left which is the 5.0.1 version check which I'll address when I rebase to latest master before the final push so I can see that constant.

You could add bw logic today already by following the instructions described at https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/Version.java#L101 I think?

markharwood commented:
I'll rebase and force push here

colings86 commented:
My worry with this idea is that it adds more complexity to an already very complex aggregation. I like the idea but I wonder if there is a way to implement it that doesn't add more complexity

markharwood commented:
>it adds more complexity to an already very complex aggregation

Do you mean the end-user complexity or the internal impl? 

Internally the aggs are untouched and it just adds another means of filtering to a class already designed to encapsulate the different means of value filtering.

markharwood commented:
> I wonder if there is a way to implement it that doesn't add more complexity

The problem is bucket explosions in a single request. The possible solutions are:
1) **Document partitioning** - doesn't work when considering > 1 doc per term e.g. finding first and last docs for a websessionID
2) **Breadth first with client-supplied Term lists** clients could filter analysis using term-exclude clauses to signal what they have already seen in previous requests but 
    * breadth-first doesn't work if top term-agg sorts terms by child aggs e.g. finding last access dates for account IDs.
    * the `exclude` term lists can get prohibitively large.
3) **Server-maintained Term lists** - similar to 2) but the server would have to maintain state in a session similar to scan/scroll to remember what had already been processed.
4) **Term partitioning** the basis of this PR
5) **Map-reduce** don't use the aggs framework at all but use big sorts and merges to analyze all the data in a way that scales with data volumes.
6) **Entity-centric indexing** aggregate data using custom update scripts and periodic updates to maintain a position e.g. each each web session's duration or last access.

jpountz commented:
As far as complexity is concerned, I agree aggs can't afford more complexity but I think this PR is not too bad since it is contained and focuses on an include impl, without adding more APIs or making existing APIs more complex.

jpountz commented on core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/support/IncludeExclude.java:
```diff
@@ -250,13 +304,30 @@ public IncludeExclude(long[] includeValues, long[] excludeValues) {
 
+    public IncludeExclude(int partition, int numPartitions) {
+        if (partition < 0 || partition >= numPartitions) {
+            throw new IllegalArgumentException(" Parition must be >=0 and < numPartition which is "+numPartitions);
```
s/Parition/Partition/
Also the leading space in the error message looks unnecessary?

jpountz commented on core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/support/IncludeExclude.java:
```diff
@@ -436,12 +518,24 @@ public boolean token(String currentFieldName, XContentParser.Token token, XConte
                         } else if (token == XContentParser.Token.VALUE_STRING) {
                             if (parseFieldMatcher.match(currentFieldName, PATTERN_FIELD)) {
                                 otherOptions.put(INCLUDE_FIELD, parser.text());
                             }
```
can you throw an exception here too? I'm worried of users providing the partition numbers as strings, which would be ignored

jpountz commented on core/src/test/java/org/elasticsearch/search/aggregations/bucket/DoubleTermsIT.java:
```diff
@@ -359,6 +362,43 @@ private void testIncludeExcludeResults(double[] includes, double[] excludes, dou
+        Terms terms = allResponse.getAggregations().get("terms");
+        assertThat(terms, notNullValue());
+        assertThat(terms.getName(), equalTo("terms"));
+        int expectedCardinality = terms.getBuckets().size();
```
can you set a very high size parameter on the terms aggregation to make sure we got all terms?

jpountz commented on core/src/test/java/org/elasticsearch/search/aggregations/bucket/DoubleTermsIT.java:
```diff
@@ -359,6 +362,43 @@ private void testIncludeExcludeResults(double[] includes, double[] excludes, dou
+            assertThat(terms.getName(), equalTo("terms"));
+            for (Bucket bucket : terms.getBuckets()) {
+                assertFalse(foundTerms.contains(bucket.getKeyAsNumber()));
+                foundTerms.add(bucket.getKeyAsNumber());
```
this can be collapsed into `assertTrue(foundTerms.add(bucket.getKeyAsNumber()))`

jpountz commented on core/src/test/java/org/elasticsearch/search/aggregations/bucket/StringTermsIT.java:
```diff
@@ -455,6 +457,38 @@ public void testSingleValueFieldWithExactTermFiltering() throws Exception {
+                }
+            }
+        }
+    }
```
no multivalued test like numerics?

jpountz commented on docs/reference/aggregations/bucket/terms-aggregation.asciidoc:
```diff
@@ -561,6 +566,67 @@ strings that represent the terms as they are found in the index:
+         }
+      }
+   }
+}
```
can you use underscore case in this example to be more consistent with the rest of the docs?

colings86 commented:
I have had a chat with @markharwood and I am ok with this be merged given the logic is almost exclusively contained within the IncludeExclude part of the terms agg.

colings86 commented on core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/support/IncludeExclude.java:
```diff
@@ -517,6 +644,11 @@ public boolean isRegexBased() {
         return include != null || exclude != null;
     }
 
+    public boolean isPartitionFilter() {
```
nit: I wonder if we should name this `isPartitionBased()` to be consistent with the `isRegexBased()`?

markharwood commented:
Ok with this @jpountz ? If so which branches shall I target?

jpountz commented on core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/TermsAggregatorFactory.java:
```diff
@@ -131,7 +131,10 @@ protected Aggregator doCreateInternal(ValuesSource valuesSource, Aggregator pare
+                // Additionally, if using partitioned terms the regular global 
+                // ordinals would be sparse so we opt for hash
+                if (Aggregator.descendsFromBucketAggregator(parent) ||
+                        (includeExclude != null && includeExclude.isPartitionBased())) {
```
++

jpountz commented on core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/support/IncludeExclude.java:
```diff
@@ -480,15 +579,43 @@ public boolean token(String currentFieldName, XContentParser.Token token, XConte
+                    if (numPartitionsObject instanceof Integer) {
+                        numPartitions = (Integer) numPartitionsObject;
+                        if (numPartitions < 2) {
+                            throw new IllegalArgumentException(PARTITION_FIELD.getPreferredName() + " must be >1");
```
s/PARTITION_FIELD/NUM_PARTITIONS_FIELD/

jpountz commented:
> If so which branches shall I target?

master and 5.x I would say, meaning we should update the version constant to be 5_2_0 since a branch for 5.1 has been cut

jpountz commented:
You might want to have a look at the changes introduced to unreleased versions at https://github.com/elastic/elasticsearch/pull/21760.

markharwood commented:
Rebased and squashed. 
We now have a Version.V_5_2_0_UNRELEASED constant in master so using that. 
Fixed the error message with the wrong field name too.
I think this is good to go now.

jpountz commented:
+1

markharwood commented:
Pushed to master https://github.com/elastic/elasticsearch/commit/aa60e5cc07ef3d87fd7fc9adc3fc481785256924 and 5.x https://github.com/elastic/elasticsearch/commit/d78ae8690d6900a5a4f31ff3e1268314f162f14a
